### ISSUE TITLE
Training on Medical Data

### ISSUE THREAD

**Issue by abdksyed** [hkchengrex/XMem#87]
Hi,

Thanks for such incredible work.

I want to train XMem on surgery data. I have videos of robotic surgery (like [CholecT80](http://camma.u-strasbg.fr/datasets)). I see that in the TRAINING.md file, there are four training stages (0: static images; 1: BL30K; 2: longer main training; 3: shorter main training). What would each stage represent for custom data like medical and surgery data?

Should I convert surgery videos and use them as static images?
What would be the best way to go for fine-tuning?

Thanks

**Comment by hkchengrex** [comment]
Static image pretraining is to kickstart training with some easy and cheap image segmentation data. I would probably start by fine-tuning from our existing models on your video data and see how well it works. The static image deformation we use can also be applied to videos if the training data is too limited. I don't know what's best -- it depends on the amount/type of data that you have.

**Comment by FJGEODEV** [comment]
> Static image pretraining is to kickstart training with some easy and cheap image segmentation data. I would probably start by fine-tuning from our existing models on your video data and see how well it works. The static image deformation we use can also be applied to videos if the training data is too limited. I don't know what's best -- it depends on the amount/type of data that you have.

For a fine-tuning job based on your pre-trained model, should I follow this:

`python -m torch.distributed.launch --master_port 25763 --nproc_per_node=2 train.py --exp_id retrain_stage3_only --stage 3 --load_network saves/XMem-s0.pth --generic_path custom_data_root --dataset G`

and my custom data should look like this:

```
├── custom_data_root
│   ├── JPEGImages
│   │   ├── video1
│   │   │   ├── 00001.jpg
│   │   │   ├── 00002.jpg
│   │   │   ├── ...
│   │   └── ...
│   ├── Annotations
│   │   ├── video1
│   │   │   ├── 00001.png
│   │   │   ├── ...
│   │   └── ...
```

Is it right? Thanks!

**Comment by FJGEODEV** [comment]
> > Static image pretraining is to kickstart training with some easy and cheap image segmentation data. I would probably start by fine-tuning from our existing models on your video data and see how well it works. The static image deformation we use can also be applied to videos if the training data is too limited. I don't know what's best -- it depends on the amount/type of data that you have.
> 
> For a fine-tuning job based on your pre-trained model, should I follow this:
> 
> `python -m torch.distributed.launch --master_port 25763 --nproc_per_node=2 train.py --exp_id retrain_stage3_only --stage 3 --load_network saves/XMem-s0.pth --generic_path custom_data_root --dataset G`
> 
> and my custom data should look like this:
> 
> ```
> ├── custom_data_root
> │   ├── JPEGImages
> │   │   ├── video1
> │   │   │   ├── 00001.jpg
> │   │   │   ├── 00002.jpg
> │   │   │   ├── ...
> │   │   └── ...
> │   ├── Annotations
> │   │   ├── video1
> │   │   │   ├── 00001.png
> │   │   │   ├── ...
> │   │   └── ...
> ```
> 
> Is it right? Thanks!

Looks like` --generic_path custom_data_root --dataset ` for eval.py only, how can I point this fine-tuning command line to my `custom_data_root`? Thanks.

**Comment by FJGEODEV** [comment]
One more question: for fine-tuning static image data, we can use this:

`python -m torch.distributed.run --master_port 25763 --nproc_per_node=2 train.py --exp_id retrain_stage3_only --stage 0 --load_network saves/XMem.pth`

Is `--exp_id retrain_stage3_only --stage 0 ` a good setting?
Here `--load_network saves/XMem.pth` should I use `XMem.pth` or `XMem_s012.pth` for the best performance?

Thanks!

**Comment by hkchengrex** [comment]
For retraining on video data, start by modifying this: https://github.com/hkchengrex/XMem/blob/5fe108138d457774f91ace362ecf97b9491f81d9/train.py#L127
For retraining on static images, look at this: https://github.com/hkchengrex/XMem/blob/5fe108138d457774f91ace362ecf97b9491f81d9/train.py#L164

`exp_id` is just an idenifier. It has no effects on training.
stage 0 is for static image pertraining and stage 3 is for video main training.
Whether XMem or XMem_s012 is better depends on your use case.

**Comment by FJGEODEV** [comment]
OK. Thanks, will try!

**Comment by FJGEODEV** [comment]
Another question: if I scribble on the predicted mask, e.g., to fill out some holes, I assume that won't change the model weight (finetune the model), just change what we want to predict, then send to the pretrained model to predict it again, is it right? Sometimes the second scribble fills that hole but another part is affected. 

Is there a way to finetune the model on the fly as we scribble the new one mask?

Thanks.

**Comment by hkchengrex** [comment]
This is a fundamental problem of a lot of the interactive image segmentation algorithms. I don't think finetuning can solve it. You can try implementing local control in [MiVOS](https://github.com/hkchengrex/MiVOS).

**Comment by FJGEODEV** [comment]
> This is a fundamental problem of a lot of the interactive image segmentation algorithms. I don't think finetuning can solve it. You can try implementing local control in [MiVOS](https://github.com/hkchengrex/MiVOS).

Thanks, good to know, appreciated your response.

**Comment by abdksyed** [comment]
Hello,

I wanted to know why we provide a few frames from a video in the VOSDataset instead of taking all the frames.

Doesn't the data loader have to give the entire video frames for the model to learn, why are we providing only 3(default in the VOSDataset definition) or 8(default in configuration.py) frames from each video?

**Comment by hkchengrex** [comment]
Training on long sequences requires back-propagation through many frames which can be memory demanding.

**Comment by abdksyed** [comment]
But what if I have a few videos, like 10, each with hundreds of frames. 
Won't fine-tuning the model with 10 videos, each 3-8 frames, not learn anything?

**Comment by hkchengrex** [comment]
@abdksyed In each epoch the dataloader picks a different subset of frames. However, this process is (kind of) uniformly random (which means the number of times a frame is picked is kind of a binomial distribution). You can modify the data loading logic if you prefer a more "even" selection over long videos.

**Comment by abdksyed** [comment]
Hello, I started the training, but the loss fluctuates a lot.

So I running only stage 3 and using only fine-tuning, which is `s3_iteration = 0` and `s3_finetuning = 50`. I have 16 videos, so with batch_size of 8, each epoch is 2 iterations, and I wanted to fine_tine for 25 epochs.

The training starts with a loss of 4.16, then in the next epoch reduces to 1.9, and then to 1.8. But in epoch 4, it increases to 3.9, and this pattern repeats. 

I tried reducing the learning rate to 1e-6 and 1e-7, but the same is happening.

Any idea, why this is happening?

**Comment by hkchengrex** [comment]
https://github.com/hkchengrex/XMem/issues/26#issuecomment-1233109935

**Comment by abdksyed** [comment]
I see, so the fluctuations are expected. But when doing testing, the results are worse than zero-shot.

I did zero shot segmentation on my data and was getting mean IoU of around 60%, but after training on my data and doing testing on that same data to see if the training was helpful, to my surprise, the IOU is even worse than zero-shot around 30-40%.

Also, since the loss function fluctuates, how to determine when to stop the training?

**Comment by hkchengrex** [comment]
On second thought, it might not be the cause of your fluctuation since you are only training for 50 iterations, which is outside of the "warmup" range (https://github.com/hkchengrex/XMem/blob/083698bbb4c5ac0ffe1a8923a6c313de46169983/util/configuration.py#L78-L79). 

It might be because we are sampling frames at random?

**Comment by abdksyed** [comment]
Yes, random sampling would be a cause for high error. I will try to train for 10,000 iterations and try to see if it will overfit or not.

Whereas the warmup, I made the start_start and end_warm as -1.
Since I am just fine-tuning on top of pre-trained XMem.pth, I didn't feel the need to do warm up the learning rate.

**Comment by hkchengrex** [comment]
It is referring to the timestep.

**Comment by abdksyed** [comment]
While doing inference, if the videos are of different size.
Like Video1 is 1280x1024 and Video2 is 960x800.

When running inference in a loop, one video after another, the inference gives error of tensor mismatch.

Is it due to the fact that the Video1 frames are in memory when running Video2, and there is a mismatch in size? 
So is it preferred to do `clear_memory` everytime we try to do inference on new video?

**Comment by hkchengrex** [comment]
In `eval.py`, the memory storage should be re-created for every new video. There should be no need to clear memory.
